securedrop/manage.py

```diff
@@ -57,56 +59,152 @@ def _stop_test_rqworker():
+        # https://www.siafoo.net/snippet/88
+        shell_colors = {
+            'gray': '30',
+            'red': '31',
+            'green': '32',
+            'yellow': '33',
+            'blue': '34',
+            'magenta': '35',
+            'cyan': '36',
+            'white': '37',
+            'crimson': '38',
+            'highlighted_red': '41',
+            'highlighted_green': '42',
+            'highlighted_brown': '43',
+            'highlighted_blue': '44',
+            'highlighted_magenta': '45',
+            'highlighted_cyan': '46',
+            'highlighted_gray': '47',
+            'highlighted_crimson': '48'
+        }
+
+        # http://stackoverflow.com/a/2330297/1093000
+        attrs = []
+        attrs.append(shell_colors[color])
+        if bold:
+            attrs.append('1')
+
+        return '\x1b[{}m{}\x1b[0m'.format(';'.join(attrs), s)
+
+
+    class DevServerProcess(subprocess.Popen):
+
+        def __init__(self, label, cmd, color):
+            self.label = label
+            self.cmd = cmd
+            self.color = color
+
+            super(DevServerProcess, self).__init__(
+                self.cmd,
+                stdin  = subprocess.PIPE,
+                stdout = subprocess.PIPE,
+                stderr = subprocess.STDOUT,
+                preexec_fn = os.setsid)
+
+        def print_label(self, to):
+            label = "\n => {} <= \n\n".format(self.label)
+            if to.isatty():
+                label = colorize(label, self.color, True)
+            to.write(label)
+
+
+    class DevServerProcessMonitor(object):
+
+        def __init__(self, procs):
+            self.procs = procs
+            self.streams = [proc.stdout for proc in procs]
+            self.last_proc = None
+            atexit.register(self.cleanup)
+
+        def _get_proc_by_stream(self, stream):
+            # TODO: this could be optimized, but it hardly matters when there
+            # are only two processes being monitored.
+            for proc in self.procs:
+                if id(stream) == id(proc.stdout):
+                    return proc
+
+        def monitor(self):
+            while True:
+                # TODO: we currently don't handle input, which makes using an
+                # interactive debugger like pdb impossible. Since Flask provides
+                # a featureful in-browser debugger, I'll accept that pdb is
+                # broken for now. If someone really wants it, they should be
+                # able to change this function to make it work (although I'm not
+                # sure how hard it would be).
+                #
+                # If you really want to use pdb, you can just run the
+                # application scripts individually (`python source.py` or
+                # `python journalist.py`).
+                rstreams, _, _ = select.select(self.streams, [], [])
+
+                for stream in rstreams:
+                    current_proc = self._get_proc_by_stream(stream)
```

Comment:
another idea would be to create a class `Procs` that just tracks the multiple processes

it could have methods like:
`procs.select()` (line 155)
`procs.any_terminated()` (line 170)
`procs.wait()` (line 183)